kernel/sched: Update runtime stats

peter-mitsis opened the pull request:
Extends the thread usage stats to allow the reporting of ...
  1. Current thread usage
  2. Peak thread usage
  3. Average thread usage
  4. Total thread usage

The tracking of thread information begins when it is first
scheduled to run and finishes when it is scheduled out.

Additionally, this feature can now be both enabled and disabled
on a per core basis at runtime.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

nordic-krch commented on subsys/shell/modules/kernel_service.c:
```diff
@@ -107,13 +107,28 @@ static void shell_tdata_dump(const struct k_thread *cthread, void *user_data)
 		shell_print(shell, "\tTotal execution cycles: %u (%u %%)",
 			    (uint32_t)rt_stats_thread.execution_cycles,
 			    pcnt);
+		shell_print(shell, "\tCurrent execution cycles: %u",
```
is there a reason why we need to distinguish that? Seems to me that this `#ifdef CONFIG_64BIT` could be removed as distinction between 64BIT and 32BIT got removed by https://github.com/zephyrproject-rtos/zephyr/commit/f169c5bc13707ad828a8ec34b3e9422f33e4861c

peter-mitsis replied:
Will look into that.

nordic-krch reviewed (commented):
Few comments. I wasn't going deep into the implementation as i'm not very familiar with stats gathering mechanism.

peter-mitsis commented on include/debug/thread_analyzer.h:
```diff
@@ -32,6 +32,10 @@ struct thread_analyzer_info {
 
 #ifdef CONFIG_THREAD_RUNTIME_STATS
 	unsigned int utilization;
+	uint64_t  current_cycles;
```
It is a good observation, and I'm was not entirely happy with the multiple definitions. Similar data exists in struct thread_analyzer_info, struct k_thread_runtime_stats, struct _thread_base and struct _cpu. The first two cases are more output/user oriented (slightly massaged data), and the latter two more internally oriented (raw data).

If desired, I am quite happy to re-jigger it so that there are two structures--one internally oriented (raw data), and one output/user oriented.

peter-mitsis commented on kernel/Kconfig:
```diff
@@ -395,6 +395,14 @@ config SCHED_THREAD_USAGE_ALL
 	  Maintain a sum of all non-idle thread cycle usage.
 
+config SCHED_THREAD_USAGE_AUTO_ENABLE
+	bool "Automatically enable runtime usage statistics at runtime"
```
You're right. I think it would probably be better stated as ...

"Automatically enable runtime usage statistics during system initialization"

andyross commented on kernel/sched.c:
```diff
@@ -1760,61 +1760,227 @@ static uint32_t usage_now(void)
-	 * synchronization as long as _usage() treats it as volatile
-	 * (we can't race with _stop() by design).
-	 */
-	_current_cpu->usage0 = usage_now();
```
So... this is a **lot** heavier than the earlier implementation, which was expected to be like 4 cycles or whatever.  Remember that this is called out of the guts of the scheduler for every context switch.  We might be in interrupt context, we might be dealing with I/O bound, rapidly-switching threads, etc...

Is there no way to defer a bunch of this work until someone tries to look at the results via z_sched_cpu_usage() or whenever?

peter-mitsis replied:
Yeah, it is much heavier. I wasn't thrilled about that.

I've not yet been able to think of a way to move more of the work out of the context switch, but I will continue to ponder.

andyross commented on include/kernel/thread.h:
```diff
@@ -171,7 +173,25 @@ struct _thread_userspace_local_data {
+	uint64_t current_cycles;      /* current # of non-idle cycles */
+	uint64_t peak_cycles;         /* peak # of non-idle cycles */
+	uint64_t total_cycles;        /* total # of non-idle cycles */
+	uint64_t average_cycles;      /* average # of non-idle cycles */
```
Seems like lots of apps will want to trace cycle usage.  Probably far fewer will want 40 bytes of extra analysis in their thread structs?  Seems like this wants to be under a CONFIG_SCHED_THREAD_USAGE_ANALYSIS or something.

peter-mitsis replied:
I like this idea.

andyross commented on kernel/sched.c:
```diff
@@ -1760,61 +1760,227 @@ static uint32_t usage_now(void)
+
+	key = k_spin_lock(&usage_lock);
+
+	if (_current_cpu->track_usage == 0) {
```
Move this test (and the other symmetric relatives) out of the lock.  The track_usage byte is only set from the current CPU (meaning it can't race against other contexts unless our local CPU gets preempted) and within a spinlock (meaning we can't be preempted).

Not having to take and release a lock on every context switch is actually big win.

andyross replied:
Sorry, to clarify: not having to take and release a lock twice on every context switch and once on every ISR is actually a big win. :)

peter-mitsis replied:
You're right. I should've seen that.

peter-mitsis replied:
Wait a minute, that claim isn't right.

There can be context switch between the test and the when the lock is taken. This can result in another thread being scheduled, which could in turn attempt to do another 'enable'.  Most of what is going to happen in this particular if-block will be inconsequential if it gets executed twice. However, the num_windows++ can get executed multiple times  resulting in a lower average calculation.

Is the resulting error in the average calculation going to be significant? Early on, perhaps. But as the system continues to execute the error can be expected to assymptotically reduce to zero.

The purist in me wants to keep the error to zero. However, the pragmatic side also recognizes that it *might* be worth the trade-off. Thoughts?

peter-mitsis replied:
Self-forehead slap ... atomic test-and-set/test-and-clear.

peter-mitsis replied:
Hmmm ... Even atomics do not quite work for what is needed.

andyross commented on include/kernel/thread.h:
```diff
@@ -171,7 +173,25 @@ struct _thread_userspace_local_data {
+	uint64_t current_cycles;      /* current # of non-idle cycles */
+	uint64_t peak_cycles;         /* peak # of non-idle cycles */
+	uint64_t total_cycles;        /* total # of non-idle cycles */
+	uint64_t average_cycles;      /* average # of non-idle cycles */
```
Also: "average_cycles" needs docs.  It looks like it's not an average over time but the "average length consumed time quantum" (i.e. on average how long does it run before it gets scheduled away).  That doesn't correspond to a number I'm familiar with from other OSes.  Might be useful, but really needs clarity.

Similarly you might consider renaming "peak" (which again tends to connote some kind of "rate" or time domain thing) to "longest" (which is what it actually stores: the longest time this thread got to run).

Finally: note that z_sched_usage_stop()/start() are called in a pair for every ISR.  So... those current/average/peak numbers are going to be polluted by whatever the interrupt behavior of the hardware devices is.  They don't really reflect thread behavior alone.  Is that intentional?

andyross replied:
Ah, OK.  Ignore the note about ISR.  I missed that you'd added an en/disable API, and that's where the statistics boundaries are placed.  That seems fine to me.

peter-mitsis replied:
Docs needed on average, and renaming peak to longest -- makes sense to me. It shall be so.

andyross reviewed (changes requested):
Sorry, just now getting around to this.  I... think this just barely gets to the level of a -1 from me.  Broadly a have a bunch of performance and bloat worries here; this is all running on hot paths and adding data to every thread struct.

Also a general point: this PR adds almost 300 lines to a sched.c file that is already pushing almost 2000.  Probably would be best to put all the thread_usage hooks and analysis into a separate file, especially since it already has its own synchronization.

andyross commented on kernel/include/ksched.h:
```diff
@@ -383,16 +383,34 @@ int z_sched_wait(struct k_spinlock *lock, k_spinlock_key_t key,
 #ifdef CONFIG_SCHED_THREAD_USAGE
 	z_sched_usage_stop();
-	z_sched_usage_start(thread);
+	z_sched_usage_start(thread, 1);
```
Style note: second argument is declared as a bool, so should be passed as "true" and not "1".  I think there's some static analysis tool or MISRA checker or something that's picky about that.

peter-mitsis replied:
Will fix.

andyross commented on include/kernel_structs.h:
```diff
@@ -131,7 +132,17 @@ struct _cpu {
+	 * [track_usage] is used to enable/disable this feature.
+	 */
+
+	uint8_t	 track_usage;
```
Hm... I guess I'm a little confused what the design here is.  I see what it does: track_usage gates the collection of thread data on a specific CPU.  I just don't see what the point is?  Why would you want to control this per-CPU and not per-thread?

If it's just to avoid performance impact on those other CPUs, then let's fix the impact (some notes elsewhere to that effect).

If it's because you only expect certain threads to run on certain CPUs, then... this seems like a footgun to me.  There is already a k_thread_cpu_mask_*() API to do thread pinning.  Duplicating it in a different way here opens up the possibility of missing data from a thread that should have been pinned but wasn't, or conversely getting a surprising zero out of your tracking because you forgot to enable the right CPU, etc...

peter-mitsis replied:
I was thinking that per thread management may lead to it being too fine-grained and thus a pain to manage from an app developer perspective, and thought that per-core might be a nice balance. As I ponder this more now, I think that I should have though it through more clearly before.

peter-mitsis replied:
I've reworked this one (will upload soon) so that there are two enable/disable pairs--one for gathering thread stats and the other for gathering the CPU stats. Flipping one should not impact the other.

andyross commented on kernel/include/ksched.h:
```diff
@@ -383,16 +383,34 @@ int z_sched_wait(struct k_spinlock *lock, k_spinlock_key_t key,
+ * @param thread Identifies thread for which to track cycles
+ * @param new_window False unless a new thread is being switched in
+ */
+void z_sched_usage_start(struct k_thread *thread, bool new_window);
```
I'm looking, and I don't see anywhere this gets called without a static "1" for the second argument?  Did something get forgotten?  I'm guessing maybe it's a parallel mixup to mine earlier, confusing the start of a thread scheduling quantum with the start of a "cpu usage window"?  As it stands, it looks like indeed every interrupt is going to split the windows, no?

peter-mitsis replied:
Great catch!  It was only supposed to get called with "1" (or "true" in one place).

Must definitely fix.

peter-mitsis replied:
I went back through my notes, and this was actually a leftover artefact from a previous idea.

A new window occurs whenever a new thread is scheduled in. Unless I am missing something, this is not something that should be occurring every interrupt (unless a new thread is getting scheduled to run after every interrupt).

andyross reviewed (commented):
A few more notes.

teburd commented on include/kernel/stats.h:
```diff
@@ -0,0 +1,24 @@
+ * and CPU usage.
+ */
+
+struct k_cycle_stats {
```
This seems like it could use the stats subsys here. It would allow exporting these with mcumgr initially and potentially more exporters in the future.

peter-mitsis replied:
I am not familiar with that subsys yet.

carlescufi replied:
I agree with @teburd. The stats subsystem can also be transferred over mcumgr after, which is very useful.

peter-mitsis replied:
I've been reading up on the stats subsys a bit, and I think this does bear looking into more. It could be a handy enhancement for a subsequent work item.

peter-mitsis committed:
kernel: move thread usage routines to own file

Moves the CONFIG_SCHED_THREAD_USAGE block of code out of sched.c
into its own file. Not only do they employ their own private
spin lock, but it is expected that additional usage routines will be
added in the future.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: update z_sched_thread_usage()

This commit does two things to the z_sched_thread_usage(). First,
it updates the API so that it accepts a pointer to the runtime
stats instead of simply returning the usage cycles. This gives it
the flexibility to retrieve additional statistics in the future.

Second, the runtime stats are only updated if the specified thread
is the current thread running on the current core.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: extend thread runtime stats

When the new Kconfig option CONFIG_SCHED_THREAD_USAGE_ANALYSIS
is enabled, additional timing stats are collected during context
switches. This extra information allows a developer to obtain the
the current, longest, average and total lengths of the time that
a thread has been scheduled to execute.

A developer can in turn use this information to tune their app and/or
alter their scheduling policies.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: extend CPU runtime stats

Extends the CPU usage runtime stats to track current, total, peak
and average usage (as bounded by the scheduling of the idle thread).
This permits a developer to obtain more system information if desired
to tune the system.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
shell: report thread runtime stats

Updates the kernel service to display the extended thread
runtime stats when the following shell command is issued.

 kernel threads

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
debug: thread analyzer to display runtime stats

Updates the thread analyzer to display the thread runtime
stats.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel/sched: enable/disable runtime stats

Adds support to enable/disable both thread and cpu runtime
stats.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

andyross commented on include/kernel.h:
```diff
@@ -5888,6 +5888,50 @@ int k_thread_runtime_stats_get(k_tid_t thread,
+ * This routine enables the gathering of CPU runtime statistics on the current
+ * core. Note that it does not affect the gathering of similar statistics for
+ * threads that are running on the core.
+ *
```
Still not sure why we want "on the current core" exposed to the API?  What would be bad if we just enforced that this function was global:  "This routine enables the gathering of runtime statistics from threads running on all CPUs".

If there's not some specific problem being solved, maybe we can leave the API (which is a promise that's hard to change) as simple as possible?

Also note some terminology glitchery: obviously "cpu" and "core" end up synonymous in all normal contexts.  Nonetheless I've tried (how successfully is left as an exercise for the reader) to be rigorous in Zephyr about using "cpu" exclusively as far as the general kernel goes.  The word "core" should ideally appear only in the platform layer where it reflects jargon that we can't control.

peter-mitsis replied:
My thinking about "on the current core" was that it may not be guaranteed that each cpu will use the same timer. As a result, to do this properly across all cores increases the complexity (IPIs to inform each core to stop and update its stats timings accordingly).

However, if this is a case that we don't want/need to worry about, then I can update this to start/stop the gathering of cpu statistics based on the cycles read from the one timer source.

andyross replied:
Ah, OK.  Yeah, timers are guaranteed consistent across cores, that's a hard requirement.  Everything in the timer subsystem breaks if you don't have cross-cpu time accuracy.

Let's take the core stuff out of the API.  It just seems like a needless complication to me.

andyross reviewed (approved):
Out of stuff to complain about, thanks.

alexsven commented:
Hello, I saw that after this PR had been merged, the CPU load seems to be off. Before this PR (tested hash 0bd3d46) I got **99%** for the idle thread in blinky. After this PR I get **49%**. I've tested this on both a nRF52840 and a nRF5340
